package/MDAnalysis/core/topology.py

```diff
@@ -324,9 +325,9 @@ def segments2residues_1d(self, six):
 
         """
         try:
-            return np.concatenate([self._SR[s] for s in six])
-        except TypeError:
-            return self._SR[six].copy()
+            return np.concatenate(self._SR[six])
+        except ValueError:  # six is not iterable, length 1 or empty
+            return self._SR[six].astype(np.int64, copy=True)
```

Comment:
@zemanj is this changing because instead of trying to iterate an integer and getting a `TypeError`, we instead pass an integer to `concatenate` and get a `ValueError`?  And then we don't need the list comprehension because we're just fancy indexing `_SR`?

The `astype` call seems unnecessary, so why not just `np.copy()` if we want to force a copy?

Reply:
The `astype()` is necessary if `six` is empty: `TransTable._SR.dtype.type` is `numpy.object_`, and the array returned by `TransTable.residues2atoms_1d()` should always be that of atom indices (even if it's empty). Unfortunately, `_SR[[]]` has the same dtype as `_SR`, so the type conversion is for that corner case.
I have to admit that the performance of `ndarray.copy()` seems to be roughly twice as fast as that of `ndarray.astype(..., copy=True)` but that doesn't really matter since the exception handling happening before the call is about a hundred times slower.

While looking at this again, I discovered that `TransTable._SR` is effectively constructed from `TransTable._AR`, which has dtype `np.intp`. Even though we probably don't really support 32 bit architectures so that `np.intp` will always end up being `np.int64`, we should be consistent here and use `np.intp`.

Reply:
> is this changing because instead of trying to iterate an integer and getting a `TypeError`, we instead pass an integer to `concatenate` and get a `ValueError`? And then we don't need the list comprehension because we're just fancy indexing `_SR`?

Exactly.

Reply:
My code comment is wrong, length 1 iterables are unproblematic.